components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/manager/FxaAccountManager.kt

```diff
@@ -357,38 +307,38 @@ open class FxaAccountManager(
      * @param reason A [SyncReason] indicating why this sync is being requested.
      * @param debounce Boolean flag indicating if this sync may be debounced (in case another sync executed recently).
      */
-    fun syncNowAsync(
+    suspend fun syncNow(
         reason: SyncReason,
         debounce: Boolean = false
-    ): Deferred<Unit> = CoroutineScope(coroutineContext).async {
-        // We may be in a state where we're not quite authenticated yet - but can auto-retry our
-        // authentication, instead.
-        // This is one of our trigger points for retrying - when a user asks us to sync.
-        // Another trigger point is the initialization flow of the account manager itself.
-        if (accountMigrationInFlight()) {
-            // Only trigger a retry attempt if the sync request was caused by a direct user action.
-            // We don't attempt to retry migration on 'SyncReason.Startup' because a retry will happen during
-            // account manager initialization if necessary anyway.
-            // If the retry attempt succeeds, sync will run as a by-product of regular account authentication flow,
-            // which is why we don't check for the result of a retry attempt here.
-            when (reason) {
-                SyncReason.User, SyncReason.EngineChange -> processQueueAsync(Event.RetryMigration).await()
-            }
-        } else {
-            // Make sure auth cache is populated before we try to sync.
-            maybeUpdateSyncAuthInfoCache()
-
-            // Access to syncManager is guarded by `this`.
-            synchronized(this) {
-                if (syncManager == null) {
-                    throw IllegalStateException(
-                        "Sync is not configured. Construct this class with a 'syncConfig' or use 'setSyncConfig'"
-                    )
+    ) = withContext(coroutineContext) {
+        when (val s = state) {
+            // Can't sync while we're still doing stuff.
+            is State.Active -> Unit
```

Comment:
IIUC this means the `syncNow` request is just dropped on the floor. Is it possible that this might cause a bad user experience where they request a "sync now" but we don't honour it because we happened to be doing something else at the time?

Reply:
That's right, we're dropping it here. FWIW, I think this can't really happen in practice - if we're in any of the "active" states, there isn't really a way for the user to request a sync. We're either starting/completing auth, logging out, migrating, etc - Fenix's UI, at least, doesn't expose "sync" button in any of these scenarios.

I'm not quite sure if having this check is necessary (it shouldn't happen), but left it here for completeness. This may as well produce a caught exception that we submit into Sentry, since it is an illegal event essentially (but, it exists outside of the state machine).

Another alternative is to make "sync" an external event, then it won't be dropped on the floor, and will only be processed when it makes sense.